pkg/reconciler/pipelinerun/resources/pipelinerunstate.go

```diff
@@ -232,31 +232,43 @@ func (state PipelineRunState) getNextTasks(candidateTasks sets.String) []*Resolv
 		if _, ok := candidateTasks[t.PipelineTask.Name]; ok {
 			if t.TaskRun == nil && t.Run == nil {
 				tasks = append(tasks, t)
-			} else { // Return any TaskRuns or Runs with remaining retries
-				var status *apis.Condition
-				var isCancelled bool
-				if t.TaskRun != nil {
-					status = t.TaskRun.Status.GetCondition(apis.ConditionSucceeded)
-					isCancelled = t.TaskRun.IsCancelled()
-					if status != nil {
-						isCancelled = isCancelled || status.Reason == v1beta1.TaskRunReasonCancelled.String()
-					}
+			}
+		}
+	}
+	tasks = append(tasks, state.getRetryableTasks(candidateTasks)...)
+	return tasks
+}
 
-				} else {
-					status = t.Run.Status.GetCondition(apis.ConditionSucceeded)
-					isCancelled = t.Run.IsCancelled()
-					if status != nil {
-						isCancelled = isCancelled || status.Reason == v1alpha1.RunReasonCancelled
-					}
+// getRetryableTasks returns a list of tasks which should be executed next when the pipelinerun is stopping, i.e.
+// a list of cancelled/failed tasks from candidateTasks which haven't exhausted their retries
+func (state PipelineRunState) getRetryableTasks(candidateTasks sets.String) []*ResolvedPipelineRunTask {
+	var tasks []*ResolvedPipelineRunTask
+	for _, t := range state {
+		if _, ok := candidateTasks[t.PipelineTask.Name]; ok {
+			var status *apis.Condition
+			var isCancelled bool
+			if t.TaskRun != nil {
+				status = t.TaskRun.Status.GetCondition(apis.ConditionSucceeded)
+				isCancelled = t.TaskRun.IsCancelled()
+				if status != nil {
+					isCancelled = isCancelled || status.Reason == v1beta1.TaskRunReasonCancelled.String()
+				}
+
+			} else if t.Run != nil {
+				status = t.Run.Status.GetCondition(apis.ConditionSucceeded)
+				isCancelled = t.Run.IsCancelled()
+				if status != nil {
+					isCancelled = isCancelled || status.Reason == v1alpha1.RunReasonCancelled
 				}
-				if status != nil && status.IsFalse() {
-					if !(isCancelled || status.Reason == ReasonConditionCheckFailed) {
-						if t.HasRemainingRetries() {
-							tasks = append(tasks, t)
-						}
+			}
+			if status != nil && status.IsFalse() {
```

Comment:
`IsFalse` performs a `nil` check [it looks like](https://github.com/tektoncd/pipeline/blob/23ffdba46dca71e30438f8ca5ab4f16d4b3bb8d7/vendor/knative.dev/pkg/apis/condition_types.go#L97-L99), so we don't need the `status != nil` I don't think?

Reply:
nice, thanks! done